File: src_python/trading/deribit/deribit.py
from __future__ import annotations

import time
from typing import Any

import requests

from src_python.core.http_client import create_http_client


class DeribitAuth:
    def __init__(self, client_id: str, client_secret: str, testnet: bool = False) -> None:
        self.client_id = client_id
        self.client_secret = client_secret
        self.base_url = "https://test.deribit.com/api/v2/" if testnet else "https://www.deribit.com/api/v2/"
        self.access_token: str | None = None
        self.refresh_token: str | None = None
        self.expiry_time = 0.0

    def _is_token_valid(self) -> bool:
        return self.access_token is not None and time.time() * 1000 < self.expiry_time - 30_000

    async def authenticate(self) -> None:
        client = create_http_client(self.base_url)
        try:
            response = await client.get(
                "public/auth",
                params={
                    "grant_type": "client_credentials",
                    "client_id": self.client_id,
                    "client_secret": self.client_secret,
                },
            )
            response.raise_for_status()
        finally:
            await client.aclose()
        data = response.json()["result"]
        self.access_token = data["access_token"]
        self.refresh_token = data["refresh_token"]
        self.expiry_time = time.time() * 1000 + int(data["expires_in"]) * 1000

    async def get_auth_header(self) -> dict[str, str]:
        if not self._is_token_valid():
            await self.authenticate()
        return {"Authorization": f"Bearer {self.access_token}"}


class DeribitOptionsTrader:
    def __init__(self, client_id: str, client_secret: str, testnet: bool = False) -> None:
        self.auth = DeribitAuth(client_id, client_secret, testnet)
        self.base_url = self.auth.base_url
        self.http = create_http_client(self.base_url)

    async def close(self) -> None:
        await self.http.aclose()

    async def _request(self, endpoint: str, params: dict[str, Any] | None = None) -> Any:
        headers = await self.auth.get_auth_header()
        response = await self.http.get(f"{self.base_url}{endpoint}", headers=headers, params=params or {})
        response.raise_for_status()
        return response.json()["result"]

    async def get_instrument(self, instrument_name: str) -> Any:
        return await self._request("public/get_instrument", {"instrument_name": instrument_name})

    async def get_order_book(self, instrument_name: str) -> Any:
        return await self._request("public/get_order_book", {"instrument_name": instrument_name})

    async def place_order(
        self,
        *,
        instrument_name: str,
        side: str,
        amount: float,
        order_type: str = "market",
        price: float | None = None,
        time_in_force: str = "fill_or_kill",
        label: str | None = None,
    ) -> Any:
        endpoint = "private/buy" if side == "buy" else "private/sell"
        payload: dict[str, Any] = {
            "instrument_name": instrument_name,
            "amount": amount,
            "type": order_type,
            "time_in_force": time_in_force,
            "label": label or f"arb-{int(time.time() * 1000)}",
        }
        if price is not None and order_type == "limit":
            payload["price"] = price

        headers = await self.auth.get_auth_header()
        url = f"{self.base_url}{endpoint}"
        print("DERIBIT ORDER:", url, payload)
        try:
            response = await self.http.get(url, headers=headers, params=payload)
            response.raise_for_status()
            return response.json()["result"]
        except requests.RequestException as exc:
            print("Network/Config Error:", exc)
            return None

    async def cancel_order(self, order_id: str) -> Any:
        return await self._request("private/cancel", {"order_id": order_id})

    async def get_open_orders(self, instrument_name: str | None = None) -> Any:
        params = {"instrument_name": instrument_name} if instrument_name else {}
        return await self._request("private/get_open_orders", params)

    async def get_positions(self) -> Any:
        return await self._request("private/get_positions")

File: src/trading/deribit/deribit.ts
import axios, { AxiosInstance } from "axios";

interface DeribitAuthResponse {
  access_token: string;
  refresh_token: string;
  expires_in: number;
  scope: string;
  state: string;
  token_type: string;
}

interface OrderResponse {
  order: {
    order_id: string;
    instrument_name: string;
    direction: string;
    price: number;
    amount: number;
    filled_amount: number;
    state: string;
  };
}

export class DeribitAuth {
  private clientId: string;
  private clientSecret: string;
  private baseUrl: string;
  private accessToken?: string;
  private refreshToken?: string;
  private expiryTime = 0;

  constructor(clientId: string, clientSecret: string, testnet = false) {
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.baseUrl = testnet
      ? "https://test.deribit.com/api/v2/"
      : "https://www.deribit.com/api/v2/";
  }

  private isTokenValid(): boolean {
    return !!this.accessToken && Date.now() < this.expiryTime - 30_000;
  }

  async authenticate(): Promise<void> {
    const res = await axios.get(`${this.baseUrl}public/auth`, {
      params: {
        grant_type: "client_credentials",
        client_id: this.clientId,
        client_secret: this.clientSecret,
      },
    });

    const data: DeribitAuthResponse = res.data.result;
    this.accessToken = data.access_token;
    this.refreshToken = data.refresh_token;
    this.expiryTime = Date.now() + data.expires_in * 1000;
  }

  async getAuthHeader(): Promise<{ Authorization: string }> {
    if (!this.isTokenValid()) {
      await this.authenticate();
    }
    return { Authorization: `Bearer ${this.accessToken}` };
  }

  getBaseUrl(): string {
    return this.baseUrl;
  }
}

export class DeribitOptionsTrader {
  private http: AxiosInstance;
  private auth: DeribitAuth;
  private baseUrl: string;

  constructor(clientId: string, clientSecret: string, testnet = false) {
    this.auth = new DeribitAuth(clientId, clientSecret, testnet);
    this.baseUrl = this.auth.getBaseUrl();
    this.http = axios.create({ timeout: 10_000 });
  }

  private async request<T>(
    endpoint: string,
    params: Record<string, any> = {}
  ): Promise<T> {
    const headers = await this.auth.getAuthHeader();
    const res = await this.http.get(`${this.baseUrl}${endpoint}`, {
      headers,
      params,
    });
    return res.data.result as T;
  }

  // === MARKET DATA ===
  async getInstrument(instrumentName: string): Promise<any> {
    return this.request("public/get_instrument", {
      instrument_name: instrumentName,
    });
  }

  async getOrderBook(instrumentName: string): Promise<any> {
    return this.request("public/get_order_book", {
      instrument_name: instrumentName,
    });
  }

  // === TRADING ===
  async placeOrder(params: {
    instrument_name: string;
    side: "buy" | "sell";
    amount: number;
    order_type?: "limit" | "market";
    price?: number;
    time_in_force?: "fill_or_kill";
    label?: string;
  }): Promise<OrderResponse> {
    const {
      instrument_name,
      side,
      amount,
      order_type = "market",
      price,
      time_in_force = "fill_or_kill",
      label,
    } = params;

    const endpoint = side === "buy" ? "private/buy" : "private/sell";

    const payload: Record<string, any> = {
      instrument_name,
      amount,
      type: order_type,
      time_in_force,
      label: label ?? `arb-${Date.now()}`,
    };

    if (price && order_type === "limit") {
      payload.price = price;
    }

    const headers = await this.auth.getAuthHeader();

    const url = `${this.baseUrl}${endpoint}`;

    console.log("DERIBIT ORDER: ", url, payload);

    try {
      const res = await this.http.get(url, {
        headers,
        params: payload,
      });
      return res.data.result as OrderResponse;
    } catch (err: any) {
      console.error("❌ Network/Config Error:", err.message);

      // throw err;
    }
  }

  async cancelOrder(orderId: string): Promise<any> {
    return this.request("private/cancel", { order_id: orderId });
  }

  async getOpenOrders(instrumentName?: string): Promise<any[]> {
    const params = instrumentName ? { instrument_name: instrumentName } : {};
    return this.request("private/get_open_orders", params);
  }

  async getPositions(): Promise<any[]> {
    return this.request("private/get_positions");
  }
}
